File: antlropt/org/perses/antlr/ast/AbstractAstVisitor.java
/*
 * Copyright (C) 2018-2020 University of Waterloo.
 *
 * This file is part of Perses.
 *
 * Perses is free software; you can redistribute it and/or modify it under the
 * terms of the GNU General Public License as published by the Free Software
 * Foundation; either version 3, or (at your option) any later version.
 *
 * Perses is distributed in the hope that it will be useful, but WITHOUT ANY
 * WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
 * FOR A PARTICULAR PURPOSE.  See the GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License along with
 * Perses; see the file LICENSE.  If not see <http://www.gnu.org/licenses/>.
 */
package org.perses.antlr.ast;

import java.util.Collection;
import java.util.Map;

public abstract class AbstractAstVisitor {
  public final void postorder(AbstractPersesRuleElement ast) {
    ast.foreachChildRuleElement(this::postorder);
    dispatch(ast);
  }

  public final void preorder(AbstractPersesAst ast) {
    dispatch(ast);
    ast.foreachChild(this::preorder);
  }

  public final void preorderGrammar(
      Collection<Map.Entry<RuleNameRegistry.RuleNameHandle, AbstractPersesRuleElement>> rules) {
    for (Map.Entry<RuleNameRegistry.RuleNameHandle, AbstractPersesRuleElement> entry : rules) {
      preorder(entry.getValue());
    }
  }

  private void dispatch(AbstractPersesAst ast) {
    switch (ast.getTag()) {
      case GRAMMAR:
        visit((PersesGrammar) ast);
        break;
      case RULE_REF:
        visit((PersesRuleReferenceAst) ast);
        break;
      case RULE_DEFINITION_LEXER:
        visit((PersesLexerRuleAst) ast);
        break;
      case RULE_DEFINITION_PARSER:
        visit((PersesParserRuleAst) ast);
        break;
      case LEXER_FRAGMENT_DEFINITION:
        visit((PersesFragmentLexerRuleAst) ast);
        break;
      case ACTION:
        visit((PersesActionAst) ast);
        break;
      case ALTERNATIVE_BLOCK:
        visit((PersesAlternativeBlockAst) ast);
        break;
      case EPSILON:
        visit((PersesEpsilonAst) ast);
        break;
      case LEXER_CHAR_SET:
        visit((PersesLexerCharSet) ast);
        break;
      case LEXER_SKIP_COMMAND:
        visit((PersesLexerSkipCommandAst) ast);
        break;
      case LEXER_CHANNEL_COMMAND:
        visit((PersesLexerChannelCommandAst) ast);
        break;
      case NOT:
        visit((PersesNotAst) ast);
        break;
      case OPTIONAL:
        visit((PersesOptionalAst) ast);
        break;
      case STAR:
        visit((PersesStarAst) ast);
        break;
      case PLUS:
        visit((PersesPlusAst) ast);
        break;
      case SEQUENCE:
        visit((PersesSequenceAst) ast);
        break;
      case TERMINAL:
        visit((PersesTerminalAst) ast);
        break;
      case TOKEN_SET:
        visit((PersesTokenSetAst) ast);
        break;
      case LEXER_RANGE_OPERATOR:
        visit((PersesRangeAst) ast);
        break;
      case ELEMENT_OPTIONS:
        visit((PersesElementOptionsAst) ast);
        break;
      default:
        throw new RuntimeException("Unhandled type: " + ast);
    }
  }

  protected abstract void visit(PersesTokenSetAst ast);

  protected abstract void visit(PersesTerminalAst ast);

  protected abstract void visit(PersesSequenceAst ast);

  protected abstract void visit(PersesPlusAst ast);

  protected abstract void visit(PersesStarAst ast);

  protected abstract void visit(PersesOptionalAst ast);

  protected abstract void visit(PersesNotAst ast);

  protected abstract void visit(PersesLexerSkipCommandAst ast);

  protected abstract void visit(PersesLexerChannelCommandAst ast);

  protected abstract void visit(PersesLexerCharSet ast);

  protected abstract void visit(PersesEpsilonAst ast);

  protected abstract void visit(PersesElementOptionsAst ast);

  protected abstract void visit(PersesAlternativeBlockAst ast);

  protected abstract void visit(PersesActionAst ast);

  protected abstract void visit(PersesFragmentLexerRuleAst ast);

  protected abstract void visit(PersesParserRuleAst ast);

  protected abstract void visit(PersesLexerRuleAst ast);

  protected abstract void visit(PersesRuleReferenceAst ast);

  protected abstract void visit(PersesGrammar ast);

  protected abstract void visit(PersesRangeAst ast);
}

File: antlropt/src/org/perses/antlr/ast/AbstractAstVisitor.kt
/*
 * Copyright (C) 2018-2024 University of Waterloo.
 *
 * This file is part of Perses.
 *
 * Perses is free software; you can redistribute it and/or modify it under the
 * terms of the GNU General Public License as published by the Free Software
 * Foundation; either version 3, or (at your option) any later version.
 *
 * Perses is distributed in the hope that it will be useful, but WITHOUT ANY
 * WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
 * FOR A PARTICULAR PURPOSE.  See the GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License along with
 * Perses; see the file LICENSE.  If not see <http://www.gnu.org/licenses/>.
 */
package org.perses.antlr.ast

import org.perses.antlr.ast.RuleNameRegistry.RuleNameHandle
import org.perses.util.exhaustive

abstract class AbstractAstVisitor {

  fun postorder(ast: AbstractPersesRuleElement) {
    ast.foreachChildRuleElement { postorder(it) }
    dispatch(ast)
  }

  fun preorder(ast: AbstractPersesAst) {
    dispatch(ast)
    ast.foreachChild { preorder(it) }
  }

  fun preorderGrammar(
    rules: Iterable<Map.Entry<RuleNameHandle, AbstractPersesRuleElement>>,
  ) {
    for ((_, value) in rules) {
      preorder(value)
    }
  }

  fun preorderGrammar(grammar: PersesGrammar) {
    grammar.flattenedAllRules
      .asSequence()
      .map(AbstractPersesRuleDefAst::body)
      .forEach { preorder(it) }
  }

  private fun dispatch(ast: AbstractPersesAst) {
    when (ast.tag) {
      AstTag.GRAMMAR -> visit(ast as PersesGrammar)
      AstTag.RULE_REF -> visit(ast as PersesRuleReferenceAst)
      AstTag.RULE_DEFINITION_LEXER -> visit(ast as PersesLexerRuleAst)
      AstTag.RULE_DEFINITION_PARSER -> visit(ast as PersesParserRuleAst)
      AstTag.RULE_DEFINITION_LEXER_FRAGMENT -> visit(ast as PersesFragmentLexerRuleAst)
      AstTag.ACTION -> visit(ast as PersesActionAst)
      AstTag.ALTERNATIVE_BLOCK -> visit(ast as PersesAlternativeBlockAst)
      AstTag.EPSILON -> visit(ast as PersesEpsilonAst)
      AstTag.LEXER_CHAR_SET -> visit(ast as PersesLexerCharSet)
      AstTag.LEXER_COMMAND -> visit(ast as PersesLexerCommandAst)
      AstTag.NOT -> visit(ast as PersesNotAst)
      AstTag.OPTIONAL -> visit(ast as PersesOptionalAst)
      AstTag.STAR -> visit(ast as PersesStarAst)
      AstTag.PLUS -> visit(ast as PersesPlusAst)
      AstTag.SEQUENCE -> visit(ast as PersesSequenceAst)
      AstTag.TERMINAL -> visit(ast as PersesTerminalAst)
      AstTag.TOKEN_SET -> visit(ast as PersesTokenSetAst)
      AstTag.LEXER_RANGE_OPERATOR -> visit(ast as PersesRangeAst)
      AstTag.RULE_ELEMENT_OPTION -> visit(ast as PersesRuleElementOption)
      AstTag.RULE_ELEMENT_LABEL -> visit(ast as PersesRuleElementLabel)
      else -> visitOthers(ast)
    }.exhaustive
  }

  protected abstract fun visit(ast: PersesTokenSetAst)
  protected abstract fun visit(ast: PersesTerminalAst)
  protected abstract fun visit(ast: PersesSequenceAst)
  protected abstract fun visit(ast: PersesPlusAst)
  protected abstract fun visit(ast: PersesStarAst)
  protected abstract fun visit(ast: PersesOptionalAst)
  protected abstract fun visit(ast: PersesNotAst)
  protected abstract fun visit(ast: PersesLexerCommandAst)
  protected abstract fun visit(ast: PersesLexerCharSet)
  protected abstract fun visit(ast: PersesEpsilonAst)
  protected abstract fun visit(ast: PersesAlternativeBlockAst)
  protected abstract fun visit(ast: PersesActionAst)
  protected abstract fun visit(ast: PersesRuleElementOption)
  protected abstract fun visit(ast: PersesFragmentLexerRuleAst)
  protected abstract fun visit(ast: PersesParserRuleAst)
  protected abstract fun visit(ast: PersesLexerRuleAst)
  protected abstract fun visit(ast: PersesRuleReferenceAst)
  protected abstract fun visit(ast: PersesGrammar)
  protected abstract fun visit(ast: PersesRuleElementLabel)
  protected abstract fun visit(ast: PersesRangeAst)

  protected open fun visitOthers(ast: AbstractPersesAst) {
    error("Unhandled type: $ast")
  }
}
